waspc/cli/src/Wasp/Cli/Command.hs

```diff
@@ -27,3 +43,88 @@ runCommand cmd = do
 -- TODO: What if we want to recognize errors in order to handle them?
 --   Should we add _commandErrorType? Should CommandError be parametrized by it, is that even possible?
 data CommandError = CommandError {_errorTitle :: !String, _errorMsg :: !String}
+
+data Requirement where
+  Requirement :: Requirable r => r -> Requirement
+
+class Typeable r => Requirable r where
+  -- | Check if the requirement is met and return a value representing that
+  -- requirement.
+  --
+  -- This function must always return a value: if the requirement is not met,
+  -- throw a 'CommandError'.
+  checkRequirement :: Command r
```

Comment:
I'm unsure about whether this type class should be exported from this module or not: doing so allows new requirables to be defined without modifying this file. Is that useful?

Also, should we move all the requirables into another module? Like `Wasp.Cli.Command.Requirables` or something? Or is leaving them in this module OK?

Reply:
I would probably extract it, just because the concept is somewhat standalone and it is now occupying very big part of this somewhat file.
I would move it to `Command/Require.hs` module. Or Requirement.

Reply:
While extracting it, I would consider writing one top comment in that module that explains the usage of `require` in a simple, approachable way. Kind of like if you are orally explaining it to somebody.

```
There are things that we need asserted in our code, and possibly also obtain 
connected value, e.g. ensuring we are in wasp project dir (+ learning what path to 
that dir is), checking if database connection is established, ... .
For stuff like that, we might want to ask the question multiple times, but we 
really need to do the actual check in the background only once. This is where 
`require` comes in play, as it has state which ensures that check is done only the 
first time, and after that it just delivers the results of last check. So now you 
can easily test for certain assumptions/requirements in any command.

For example, you can check if we are in wasp project root dir by doing:
EXAMPLE

All the current requirements are implemented in this module. To implement a new 
create your new type and make it an instance of Requirable, and that is it.
```

Something like this, with this level of simplicity / verbosity.

Reply:
Unfortunately there is no good way to completely remove the requirement code from `Command.hs`. Since the requirables depend on the Command monad, and the Command monad depends on Requirement, they can't be fully separated. I went for leaving the core requirement code (Requirement type, Requirable class, and require function) in `Command.hs` and moved the instances of requirable to a new module.